### DISCUSSION TITLE
basePath in the NextRequest?

### DISCUSSION THREAD

**Discussion by mildfuzz** [vercel/next.js#70280] [Help]
### Summary

I have a multi-application stack, with several applications running on custom basePath's 

I want to write a single function that will redirect a user from one to the other, but at the moment I have having to hard code the route back because I can't deduce the `basePath` from the request. 

I have logged out all the values in `request.nextUrl` and it doesn't appear to be in there, is this by design?

**Comment by Netail** [comment]
`request.nextUrl.basePath` returns the basePath

**Reply by plaa** [reply]
I am trying to find out the same, and `basePath` is empty even though it's configured in `next.config.js`.

In `next.config.js` I have `basePath: '/mybase'`, and I have `src/api/endpoint/route.ts` with the following:

```
import { NextRequest } from 'next/server';

export function GET(request: NextRequest) {
  console.log('url=', request.url);
  console.log('basePath=', request.nextUrl.basePath);
  console.log('pathname=', request.nextUrl.pathname);
  console.log('nextUrl=', request.nextUrl);
  return Response.json({ status: 'ok' }, { status: 200 });
}
```

When I make a request to `http://localhost:3100/mybase/api/endpoint?foo` the `mybase` is hidden from everything (even the absolute URLs) and `basePath` is empty:

```
url= http://localhost:3100/api/endpoint?foo
basePath=
pathname= /api/endpoint
nextUrl= NextURL {
  [Symbol(NextURLInternal)]: {
    url: URL {
      href: 'http://localhost:3100/api/endpoint?foo',
      origin: 'http://localhost:3100',
      protocol: 'http:',
      username: '',
      password: '',
      host: 'localhost:3100',
      hostname: 'localhost',
      port: '3100',
      pathname: '/api/endpoint',
      search: '?foo',
      searchParams: URLSearchParams { 'foo' => '' },
      hash: ''
    },
    options: { headers: [Object], nextConfig: undefined },
    basePath: '',
    domainLocale: undefined,
    defaultLocale: undefined,
    buildId: undefined,
    locale: undefined,
    trailingSlash: false
  }
}
```

Did you @mildfuzz ever figure out a way to reliably obtain the base path?

**Reply by plaa** [reply]
This may be related, though under different circumstances: https://github.com/vercel/next.js/issues/49883

**Comment by Kavindi487** [comment]
Hi @mildfuzz 👋

You’re right — the basePath isn’t exposed directly on the NextRequest object by design. It’s handled internally by the Next.js router layer and stripped out before the request reaches middleware or API routes.

However, there are a couple of reliable ways to work around this depending on your setup:

✅ Option 1: Use next.config.js and Import It

Since the basePath is static and known at build time, you can import it where you need it:

// next.config.js
module.exports = {
  basePath: "/app1",
};

// middleware.js
import { NextResponse } from "next/server";
import config from "../next.config.js";

export function middleware(req) {
  const base = config.basePath || "";
  return NextResponse.redirect(new URL(`${base}/dashboard`, req.url));
}


✅ Works well if you’re building each app independently.

🧠 Option 2: Derive it Dynamically from the URL

If you’re running multiple Next.js apps under different base paths (e.g. /app1, /app2, etc.), you can infer it from req.nextUrl.pathname:

const { pathname, origin } = req.nextUrl;
const baseMatch = pathname.match(/^\/(app1|app2|app3)/);
const basePath = baseMatch ? `/${baseMatch[1]}` : "";

return NextResponse.redirect(new URL(`${basePath}/dashboard`, origin));


This lets your redirect logic automatically adapt to whichever app the request came from — without hardcoding each path.

⚙️ Why It’s Not Included

Next.js removes the basePath before building the request context to ensure all internal routing stays consistent.
Including it in NextRequest could cause mismatched routes if you rewrote paths dynamically.

💬 Summary

basePath is not exposed intentionally.

You can import it from next.config.js or detect it dynamically from the URL.

This allows middleware-based redirects to work across multi-app environments safely.

Hope this helps you keep your redirect logic DRY and dynamic! 🚀
(Tested in Next.js 15.3 with App Router + Middleware.)
